apps/balancer-tools/src/app/apr/(utils)/getBlockNumberByTimestamp.ts

```diff
@@ -2,52 +2,61 @@ import { NetworkChainId } from "@bleu-balancer-tools/utils";
 
 import { blocks } from "#/lib/gql/server";
 
-const MINUTE_IN_SECONDS = 60;
+import { dateToEpoch } from "../api/(utils)/date";
 
 const GNOSIS_API_EXPLORER = (timestamp: number): string =>
   `https://api.gnosisscan.io/api?module=block&action=getblocknobytime&timestamp=${timestamp}&closest=before`;
 
 async function bestGuess(chain: number, timestamp: number): Promise<number> {
   if (chain !== 100) {
-    return 0;
+    throw new Error(
+      `Invalid chain ID for bestGuess function. Chain ID: ${chain}`,
+    );
   }
+
   const url = GNOSIS_API_EXPLORER(timestamp);
   const response = await fetch(url);
+
+  if (!response.ok) {
+    throw new Error(
+      `Failed to fetch data from Gnosis API. HTTP Status: ${response.status}, URL: ${url}`,
+    );
+  }
+
   const data = await response.json();
-  return parseInt(data.result);
-}
 
-function getTimestamps(time: Date): Record<string, string> {
-  const timestamp_gte = (time.getTime() / 1000).toString();
-  const timestamp_lt = (
-    time.getTime() / 1000 +
-    MINUTE_IN_SECONDS * 10
-  ).toString();
-  return { timestamp_gte, timestamp_lt };
+  if (!data.result) {
+    throw new Error(
+      `No block found using the Gnosis API. Timestamp: ${timestamp}`,
+    );
+  }
+
+  return Number(data.result);
 }
 
 export default async function getBlockNumberByTimestamp(
   chain: number,
-  endTime: Date,
+  endTime: number,
 ): Promise<number> {
-  if (endTime > new Date()) {
-    return -1;
+  if (endTime > dateToEpoch(new Date())) {
+    throw new Error(
+      `The specified endTime cannot be in the future. EndTime: ${endTime}`,
+    );
   }
 
-  const timestamps = getTimestamps(endTime);
-
   if (chain === NetworkChainId.GNOSIS) {
-    return bestGuess(chain, parseInt(timestamps.timestamp_gte));
+    return bestGuess(chain, endTime);
   }
 
   const data = await blocks.gql(String(chain)).Blocks({
-    timestamp_gte: timestamps.timestamp_gte,
-    timestamp_lt: timestamps.timestamp_lt,
+    timestamp_gte: endTime,
   });
 
-  if (data.blocks.length > 0) {
-    return parseInt(data.blocks[0].number);
+  if (data.blocks.length === 0) {
+    throw new Error(
+      `No blocks found in the specified time range. Chain: ${chain}, endTime: ${endTime}`,
+    );
```

Comment:
throw to avoid returning a bogus result which will only break when we do a time travel query on subgraph/contracts

Reply:
I think this pattern of throwing an exception and handling it on its caller its a better pattern, do you know if Sentry has its trace like it does in python (displaying all variables on memory) or do we have to add it ourselves?